perfkitbenchmarker/benchmarks/cassandra_stress_benchmark.py

```diff
@@ -154,6 +212,92 @@ def CheckPrerequisites():
     data.ResourcePath(FLAGS.cassandra_stress_profile)
 
 
+def CheckMetadata(metadata):
+  """Verify that metadata is valid.
+
+  Args:
+    metadata: dict. Contains metadata for this benchmark.
+  """
+  if metadata['command'] in PRELOAD_REQUIRED:
+    if metadata['pop_size'] > metadata['num_preload_keys']:
+      raise errors.Benchmarks.PrepareException(
+          'For %s modes, number of preloaded keys must be larger than or '
+          'equal to population size.', PRELOAD_REQUIRED)
+
+
+def SetupMetadata(benchmark_spec):
+  """Setup metadata.
+
+  Args:
+    benchmark_spec: The benchmark specification. Contains all data that is
+        required to run the benchmark.
+
+  Returns:
+    dist. Contains metadata for this benchmark.
+  """
+  vm_dict = benchmark_spec.vm_groups
+  metadata = {}
+  if not FLAGS.num_keys:
+    metadata['num_keys'] = (
+        NUM_KEYS_PER_CORE * vm_dict[CASSANDRA_GROUP][0].num_cpus)
+  else:
+    metadata['num_keys'] = FLAGS.num_keys
+
+  if (FLAGS.cassandra_stress_command in PRELOAD_REQUIRED and
+      not FLAGS.cassandra_stress_preload_num_keys):
+    metadata['num_preload_keys'] = metadata['num_keys']
+  else:
+    metadata['num_preload_keys'] = FLAGS.cassandra_stress_preload_num_keys
+
+  metadata.update({
+      'num_data_nodes': len(vm_dict[CASSANDRA_GROUP]),
+      'num_loader_nodes': len(vm_dict[CLIENT_GROUP]),
+      'num_cassandra_stress_threads': FLAGS.num_cassandra_stress_threads,
+      'command': FLAGS.cassandra_stress_command,
+      'consistency_level': FLAGS.cassandra_stress_consistency_level,
+      'retries': FLAGS.cassandra_stress_retries,
+      'pop_size': (FLAGS.cassandra_stress_pop_size or
+                   max(metadata['num_keys'],
+                       metadata['num_preload_keys'])),
+      'pop_dist': FLAGS.cassandra_stress_pop_distribution,
+      'pop_parameters': ','.join(FLAGS.cassandra_stress_pop_parameters)})
+
+  if FLAGS.cassandra_stress_command == USER_COMMAND:
+    metadata.update({
+        'profile': FLAGS.cassandra_stress_profile,
+        'ops': FLAGS.cassandra_stress_ops})
+  else:
+    if FLAGS.cassandra_stress_command == MIXED_COMMAND:
+      metadata['mixed_ratio'] = FLAGS.cassandra_stress_mixed_ratio
+    metadata['replication_factor'] = FLAGS.cassandra_stress_replication_factor
+  logging.info('Metadata: %s', metadata)
+  return metadata
+
+
+def PreloadCassandraServer(benchmark_spec, metadata):
+  """Preload cassandra cluster if necessary.
+
+  Args:
+    benchmark_spec: The benchmark specification. Contains all data that is
+        required to run the benchmark.
+    metadata: dict. Contains metadata for this benchmark.
+  """
+  if metadata['num_preload_keys']:
```

Comment:
Nit: I think this would be cleaner if PreloadCassandraServer always pre-loaded the database, and checking whether preload should happen was done in the caller.